src/clj/puppetlabs/puppetserver/certificate_authority.clj

```diff
@@ -747,9 +766,39 @@
   should be copied to."
   ([cacrl :- schema/Str
     localcacrl :- schema/Str]
-    (when (fs/exists? cacrl)
-      (ks/mkdirs! (fs/parent localcacrl))
-      (fs/copy cacrl localcacrl))))
+    (when (not= cacrl localcacrl)
+      (let [max-attempts 25]
+        (when (and (fs/exists? cacrl))
+          (ks/mkdirs! (fs/parent localcacrl))
+          ;; Make up to 'max-attempts' tries to copy the cacrl file to the
+          ;; localcacrl file. The content of the cacrl file is read into memory
+          ;; and parsed for valid CRL pem content during each attempt. The content
+          ;; in memory is written to the localcacrl file only if it is valid. This
+          ;; validation is done to protect against a partially written (which could
+          ;; happen if an asynchronous revocation is in progress) or currupt cacrl
+          ;; file being copied.
+          (loop [attempts-left max-attempts
+                 ca-crl-last-modified (fs/mod-time cacrl)]
+            (let [cacrl-as-string (slurp cacrl)]
+              (when-let [write-exception (write-local-cacrl!
+                                           localcacrl
+                                           cacrl-as-string)]
+                (if (zero? attempts-left)
+                  (log/error (format "%s\n%s\n%s"
+                                     (i18n/trs
+                                       "Unable to synchronize crl file {0} to {1}: {2}"
+                                       cacrl localcacrl (.getMessage write-exception))
+                                     (i18n/trs
+                                       "Recent changes to the CRL may not have taken effect.")
+                                     (i18n/trs
+                                       "To load the updated CRL, reload or restart the Puppet Server service.")))
+                  (do
+                    (Thread/sleep 100)
+                    (let [new-ca-crl-last-modified (fs/mod-time cacrl)]
+                      (recur (if (= ca-crl-last-modified new-ca-crl-last-modified)
```

Comment:
Two things, 
This seems dangerous. If, in attempting to write the CRL file out, it updates the file's modified time but still throws an IOException then this will loop forever. I think I'd prefer to keep this the same as the method in 2.x.

Secondly, if we did do it this way, I think we should assign a local variable for the output of the conditional have the recur be the last thing.

Reply:
Thanks, looks like I missed a spot where we updated that in 2.x as well. Fixing.